django/db/models/functions/text.py

```diff
@@ -65,83 +65,49 @@ def as_sqlite(self, compiler, connection, **extra_context):
+        if len(expressions) < 2:
+            raise ValueError("Concat must take at least two expressions")
+        super().__init__(*expressions, **extra)
+
+    def as_mysql(self, compiler, connection):
+        # MySQL || operator is an alias for OR unless PIPES_AS_CONCAT is enabled.
+        return Func(Value(""), *self.source_expressions, function="CONCAT_WS").as_sql(
+            compiler, connection
         )
 
-    def as_postgresql(self, compiler, connection, **extra_context):
+    def as_sql(self, compiler, connection, **extra_context):
         copy = self.copy()
         copy.set_source_expressions(
             [
-                Cast(expression, TextField())
+                Coalesce(
+                    (
+                        expression
+                        if isinstance(expression.output_field, (CharField, TextField))
+                        else Cast(expression, TextField())
+                    ),
+                    Value(""),
+                )
                 for expression in copy.get_source_expressions()
             ]
         )
-        return super(ConcatPair, copy).as_sql(
-            compiler,
-            connection,
-            **extra_context,
-        )
-
-    def as_mysql(self, compiler, connection, **extra_context):
-        # Use CONCAT_WS with an empty separator so that NULLs are ignored.
-        return super().as_sql(
-            compiler,
-            connection,
-            function="CONCAT_WS",
-            template="%(function)s('', %(expressions)s)",
-            **extra_context,
-        )
-
-    def coalesce(self):
-        # null on either side results in null for expression, wrap with coalesce
-        c = self.copy()
-        c.set_source_expressions(
-            [
-                Coalesce(expression, Value(""))
-                for expression in c.get_source_expressions()
-            ]
-        )
-        return c
+        return super(Concat, copy).as_sql(compiler, connection, **extra_context)
 
 
-class Concat(Func):
-    """
-    Concatenate text fields together. Backends that result in an entire
-    null expression when any arguments are null will wrap each argument in
-    coalesce functions to ensure a non-null result.
-    """
-
-    function = None
-    template = "%(expressions)s"
-
-    def __init__(self, *expressions, **extra):
-        if len(expressions) < 2:
-            raise ValueError("Concat must take at least two expressions")
-        paired = self._paired(expressions)
-        super().__init__(paired, **extra)
-
-    def _paired(self, expressions):
-        # wrap pairs of expressions in successive concat functions
-        # exp = [a, b, c, d]
-        # -> ConcatPair(a, ConcatPair(b, ConcatPair(c, d))))
-        if len(expressions) == 2:
-            return ConcatPair(*expressions)
-        return ConcatPair(expressions[0], self._paired(expressions[1:]))
+# XXX: Maybe we should just mark this thing for deprecation?
+class ConcatPair(Concat):
```

Comment:
With `||` used on SQLite, Postgres, and Oracle and `CONCAT_WS` used on MySQL the implementation detail that `ConcatPair` is completely obsolete.